File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/FunctionNestingDepth.vb
Imports System

Namespace Tests.Diagnostics
    Public Class FunctionNestingDepth

        Public Sub New()
            Do While True
                Try
                    While True
                        If (True) Then  ' Noncompliant {{Refactor this code to not nest more than 3 control flow statements.}}
'                       ^^
                            Console.Write()
                        End If
                    End While

                Catch

                End Try
            Loop
        End Sub

        Protected Overrides Sub Finalize()
            Do While True
                Try
                    While True
                        For Each x In {1, 2, 3} ' Noncompliant {{Refactor this code to not nest more than 3 control flow statements.}}
                        Next
                    End While
                Catch
                End Try
            Loop
        End Sub

        Public Sub M1()
            If True Then
                While True
                    Try
                    Catch
                    End Try
                End While

                If (True) Then

                ElseIf (True) Then
                    If (True) Then
                        Console.Write()
                    End If

                ElseIf (True) Then
                    Do
                        Do ' Noncompliant
                        Loop
                    Loop

                ElseIf (True) Then
                End If
            End If
        End Sub

        Public Sub M2()
            If True Then
                If (True) Then
                    If (True) Then
                        For i As Integer = 0 To 1 Step 0 ' Noncompliant

                        Next
                    End If
                End If
            End If
        End Sub

        Public Function M3() As Integer
            Return (New Func(Of Integer)(Function() As Integer
                                             If (True) Then
                                                 If (True) Then
                                                     If (True) Then

                                                         If (True) Then ' Noncompliant
                                                             Return 0
                                                         End If
                                                     End If
                                                 End If
                                             End If

                                             Return 42

                                         End Function))()
        End Function


        Public Property MyProperty As Integer
            Get
                If (True) Then
                    If (True) Then
                        If (True) Then

                            If (True) Then ' Noncompliant
                                Return 0
                            End If
                        End If
                    End If
                End If

                Return 42
            End Get

            Set
                If (True) Then
                    If (True) Then
                        If (True) Then

                            If (True) Then ' Noncompliant
                            End If
                        End If
                    End If
                End If
            End Set
        End Property

        Public Custom Event OnSomething As EventHandler
            AddHandler()
                If (True) Then
                    If (True) Then
                        If (True) Then

                            If (True) Then ' Noncompliant
                            End If
                        End If
                    End If
                End If
            End AddHandler
            RemoveHandler()
                If (True) Then
                    If (True) Then
                        If (True) Then

                            If (True) Then ' Noncompliant
                            End If
                        End If
                    End If
                End If
            End RemoveHandler
        End Event

        Public Shared Operator +(ByVal h1 As FunctionNestingDepth,
                             ByVal h2 As FunctionNestingDepth) As FunctionNestingDepth
            If (True) Then
                If (True) Then
                    If (True) Then

                        If (True) Then ' Noncompliant
                        End If
                    End If
                End If
            End If
            Return New FunctionNestingDepth
        End Operator

        Public Sub M5()
            If True Then
                If (True) Then
                    If (True) Then
                        For i As Integer = 0 To 1 Step 0 ' Noncompliant
                            If (True) Then ' Compliant

                            End If
                        Next
                    End If
                End If
            End If
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/FunctionNestingDepth.cs
using System;

namespace Tests.Diagnostics
{
    public class FunctionNestingDepth
    {
        public FunctionNestingDepth()
        {
            do
            {
                try
                {
                    while (true)
                    {
                        if (true) { } // Noncompliant {{Refactor this code to not nest more than 3 control flow statements.}}
//                      ^^
                    }
                }
                catch { }
            } while (true);
        }

        ~FunctionNestingDepth()
        {
            do
            {
                try
                {
                    while (true)
                    {
                        foreach (var i in new string[] { }) { } // Noncompliant
                    }
                }
                catch { }
            } while (true);
        }

        public void M1()
        {
            {
                if (true)
                {
                    while (true)
                    {
                        try { }
                        catch { }
                    }
                }
                if (true) { }
                else if (true)
                {
                    if (true) { }
                    else if (true)
                    {
                        do
                        {
                            do // Noncompliant
                            {

                            }
                            while (true);
                        }
                        while (true);
                    }
                }
                else if (true) { }
            }
        }

        public void M2()
        {
            if (true)
            {
                if (true)
                {
                    if (true)
                    {
                        for (; ; ) // Noncompliant
                        {
                        }
                    }
                }
            }
        }

        public int M3() => (new Func<int>(() =>
        {
            if (true)
                if (true)
                    if (true)
                        if (true) // Noncompliant
                            return 0;
            return 42;
        }))();

        public int M4() => (new Func<int>(() =>
        {
            if (true)
                if (true)
                    if (true)
                        return 0;
            return 42;
        }))();

        public void SwitchStatement()
        {
            if (true)
            {
                switch (42)
                {
                    case 1:
                        break;
                    case 2:
                        switch (42)
                        {
                            case 0:
                                break;
                            default:
                                break;
                        }
                }
            }
            if (true)
            {
                switch (42)
                {
                    case 1:
                        break;
                    case 2:
                        switch (42)
                        {
                            case 0:
                                if (true) { }   // Noncompliant
                                break;
                            default:
                                if (true) { }   // Noncompliant
                                break;
                        }
                }
            }
        }

        public int MyProperty
        {
            get
            {
                if (true)
                {
                    if (true)
                    {
                        if (true)
                        {
                            if (true) { } // Noncompliant
                        }
                    }
                }
                return 0;
            }
            set
            {
                if (true)
                {
                    if (true)
                    {
                        if (true)
                        {
                            if (true) { } // Noncompliant
                        }
                    }
                }
            }
        }

        public event EventHandler OnSomething
        {
            add
            {
                if (true)
                {
                    if (true)
                    {
                        if (true)
                        {
                            if (true) { } // Noncompliant
                        }
                    }
                }
            }
            remove
            {
                if (true)
                {
                    if (true)
                    {
                        if (true)
                        {
                            if (true) { } // Noncompliant
                        }
                    }
                }
            }
        }

        public static FunctionNestingDepth operator +(FunctionNestingDepth a)
        {
            if (true)
            {
                if (true)
                {
                    if (true)
                    {
                        if (true) { } // Noncompliant
                    }
                }
            }
            return null;
        }

        public static void M5()
        {
            if (true)
            {
                if (true)
                {
                    if (true)
                    {
                        if (true)  // Noncompliant
                        {
                            if (true) // Compliant
                            {

                            }
                        }
                    }
                }
            }
        }
    }
}
